parse url to set parameters and return parameters as dict

ghukill opened the pull request:
Not sure at all if this is in scope for the IIIFImageClient, but thought it was worth throwing out there.

Would be two additions: 
* ability to provide and parse a URL to set the `api_endpoint`, `image_id`, `region`, `size`, `rotation`, `quality`, and `format` of the client instance
* return a dictionary that parses the more complex parameters of `region`, `size`, and `rotation` 

We have a use case where we want to determine the height and/or width that a URL is requesting, and whether or not it's mirrored.  These two additions make that pretty easy given a URL.

e.g.
```
from piffle.iiif import IIIFImageClient
ic = IIIFImageClient()

ic.parse_from_url('http://localhost/loris/foo:bar/full/full/0/default.jpg')
In [5]: ic.api_params_as_dict()
Out[5]: 
{'region': {'full': True,
  'h': None,
  'pct': False,
  'w': None,
  'x': None,
  'y': None},
 'rotation': {'degrees': 0, 'mirrored': False},
 'size': {'exact': False, 'full': True, 'h': None, 'pct': False, 'w': None}}

# with more parameters
ic.parse_from_url('http://localhost/loris/foo:bar/2560,2560,256,256/256,/!90/default.jpg')
In [11]: ic.api_params_as_dict()
Out[11]: 
{'region': {'full': False,
  'h': '256',
  'pct': False,
  'w': '256',
  'x': '2560',
  'y': '2560'},
 'rotation': {'degrees': 90, 'mirrored': True},
 'size': {'exact': False, 'full': False, 'h': None, 'pct': False, 'w': 256}}
```

Either way, thanks for breaking this into it's own package!  Mighty handy.

rlskoeser commented on piffle/iiif.py:
```diff
@@ -108,3 +108,125 @@ def format(self, image_format):
         img.image_options['format'] = image_format
         return img
+
+    def parse_from_url(self, url):
```
This seems pretty slick -  it lets you can initialize an image client instance from a url?  I think maybe it would be a bit cleaner to make it a classmethod and return a new instance.  And what do you think of calling it `init_from_url` instead of parse?

rlskoeser commented on piffle/iiif.py:
```diff
@@ -108,3 +108,125 @@ def format(self, image_format):
+        
+        url_components = url.split('/')
+        
+        _quality, _format = url_components[-1].split('.')
```
Probably need some error handling here (and/or maybe in the init method).  What if the url is malformed somehow?  Also, what about supporting the info.json url to initialize the api endpoint and image id?

rlskoeser commented on piffle/iiif.py:
```diff
@@ -108,3 +108,125 @@ def format(self, image_format):
+        self.__init__(api_endpoint=_api_endpoint, image_id=_image_id, region=_region,
+                     size=_size, rotation=_rotation, quality=_quality, format=_format)
+                     
+    def api_params_as_dict(self):
```
Is there a clearer/shorter name we could use? dict_opts ? 

Also, I think it might be helpful to add a little more to the description to elaborate on why this is useful (i.e., to allow logical decision making based on the image features).

rlskoeser commented on piffle/iiif.py:
```diff
@@ -108,3 +108,125 @@ def format(self, image_format):
+        }
+
+    # methods to derive python dictionaries from IIIF strings
+    def _region_as_dict(self):
```
Any reason the sub dict methods shouldn't be exposed publicly? Seems like they could be useful.

rlskoeser commented on piffle/iiif.py:
```diff
@@ -108,3 +108,125 @@ def format(self, image_format):
+            region = region.split("pct:")[1]
+
+        # split to dictionary
+        region_d['x'],region_d['y'],region_d['w'],region_d['h'] = region.split(",")
```
Should these be ints, for consistency with size?

rlskoeser commented on tests/test_piffle.py:
```diff
@@ -56,3 +56,13 @@ def test_format(self):
+    def test_url_parse(self):
+        img = iiif.IIIFImageClient()
+        # simple API parameters
+        img.parse_from_url('http://imgserver.co/loris/img1/full/full/0/default.jpg')
```
You're not actually testing anything except that the method doesn't error, right?  Please add checks to confirm that the `img` object is initialized with the correct values in both cases, add checks for malformed or invalid urls (raises expected error, or whatever), and also test for info.json url if you decide to support that (or as an error case).

rlskoeser commented on tests/test_piffle.py:
```diff
@@ -56,3 +56,13 @@ def test_format(self):
+        img.api_params_as_dict()
+        # more complex API parameters
+        img.parse_from_url('http://imgserver.co/loris/img1/2560,2560,256,256/256,/!90/default.jpg')
+        img.api_params_as_dict()
```
I would break this out into a separate unit test; probably easiest to test the submethods individually and just check that you get the expected values and types for the variety of options that are allowed for each.

ghukill commented:
All sound like excellent ideas, thanks for the feedback.  Will give those changes a go and update this pull request.

rlskoeser commented:
All the updates look good.  Given that the last commit is "pre-tests" is it fair to assume more is coming on the unit tests?

Also, might be good to declare and use a custom exception class, just to make it easier to catch and identify.

ghukill commented:
Alrighty, think I've addressed most of your suggestions:

- renamed the method to `init_from_url()`, and has it return new class instance
- error handling for malformed urls, raises custom `InitURLError` exception
- used `dict_opts` for aggregated parsed parameters
- sub-methods are public (as they should have been in the first place)
- tightened up value types that are returned from parsed parameters, returning `float` when percentage is used per the spec
- added tests that assert successful init from urls, exceptions catching bag ones, and for a complex image request url, the sub-methods for parsing parameters are returning the expected dictionaries

Thanks again for taking the time to look over the initial PR and provide feedback, it was pretty rough around the edges the first go.

rlskoeser commented:
Updates look good.  I'll try to get it merged in and maybe make another release soon.

ghukill commented:
Awesome, thanks.

rlskoeser commented:
@ghukill I'm doing some clean up and extending the unit tests based on your additions, and I'm wondering if it would make more sense to store the options internally as dictionaries rather than strings.  That would mean options provided at init would be parsed and checked (which they currently aren't), and seems like it might be a more logical approach in general.  Thoughts?

ghukill commented:
That's a great idea.  I had originally envisioned methods like `region_as_dict`, `size_as_dict`, etc. to parse dictionaries from strings because they were stored as strings in the original init.  But yeah, reworking init to store the image request parameters as dictionaries right from the get-go, that makes sense.  Then you can check the values at that time, and know right away if it's a valid IIIF url.  Let me know if I can help at all.

rlskoeser commented:
I'm going to go ahead and close this pull request, since the code is merged into develop and documented in the change log.  Thanks for the contribution!
